#166214113 Pagination for book

Nennyfills opened the pull request:
#### What does this PR do?
Add  Book Pagination functionality

#### Description of Task to be completed?
- Create new endpoints for Book Pagination for all users
- Add faker library to the seeders book file
- Write unit test to test all routers

#### How should this be manually tested?
- Clone the project 
- Install all dependencies
- Run test
OR
- Run npm run db:rollback
- Start server and test manually with npm run start:dev
- Open postman use:
      1. `http://localhost:{PORT}/api/v1/books`
      2. `http://localhost:{PORT}/api/v1/books?page=1&limit=10`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166214113](https://www.pivotaltracker.com/story/show/166214113)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/45515619/59678185-9b336e00-91c3-11e9-8899-c6493a3157a8.png)
![image](https://user-images.githubusercontent.com/45515619/59678237-c6b65880-91c3-11e9-910b-15818a9dbd3e.png)
![image](https://user-images.githubusercontent.com/45515619/59678286-e6e61780-91c3-11e9-8401-7b1d0a58292f.png)
![image](https://user-images.githubusercontent.com/45515619/59678308-fb2a1480-91c3-11e9-89f4-0b5859572e34.png)


#### Questions:
N/A

henryade commented on server/controllers/BookController.js:
```diff
@@ -42,6 +42,33 @@ class BookController {
+* @param {Object} response
+* @returns {JSON} object
+*/
+  static async pagination(request, response) {
```
Class name should be capitalized

Nennyfills replied:
that is not a class is a static method in a class

henryade commented on server/controllers/BookController.js:
```diff
@@ -42,6 +42,33 @@ class BookController {
+    const { page, limit } = request.query;
+    let pageLimit = limit;
+    let offset = 0;
+    pageLimit = +pageLimit ? pageLimit : pageLimit = 10;
```
I don't understand this line

Nennyfills replied:
to have a default limit if none is given

henryade replied:
you have two equal signs and what is the plus sign for?

Nennyfills replied:
i will request you study the code well,

henryade commented on server/controllers/BookController.js:
```diff
@@ -42,6 +42,33 @@ class BookController {
+    try {
+      const countResults = await Book.findAndCountAll();
+      const { count } = countResults;
+      const pages = Math.ceil(countResults.count / pageLimit);
```
You already destructed count from countResults

Nennyfills replied:
okay

henryade commented on server/database/seeders/20190614113741-book.js:
```diff
@@ -0,0 +1,115 @@
+/* eslint-disable */
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('Books', [{
+    title: 'Take me home',
```
Did you use faker here

Nennyfills replied:
added

henryade commented on server/routes/auth.js:
```diff
@@ -16,5 +19,7 @@ auth.post(`${BASE_URL}/signup`, signUpValidation(), createUser);
 // Verify User route
 auth.get(`${BASE_URL}/verify`, verifyToken, verifyUser);
+// pagination
+auth.get(`${BASE_URL}`, pagination);
```
This is in the wrong file. This has nothing to do with auth.js

Nennyfills replied:
ok

Adekoreday replied:
it might be in the user route

henryade commented on test/server/routes/auth.test.js:
```diff
@@ -406,6 +406,27 @@ describe('AUTH', () => {
+          expect(response.body).to.be.an('object');
+          expect(response.body.message).to.equal('request succefully');
+          done();
+        });
```
No test for failure?

henryade reviewed (changes requested):
No swagger docs?

Lundii reviewed (dismissed):
Good job. I think 
* you should add some validations for the query parameters  should incase a wrong query is given example a string
* with respect to above, write test for wrong query parameters
* work on the changes observed by Adeola especially the location of your route

Nennyfills commented:
> Good job. I think
> 
> * you should add some validations for the query parameters  should incase a wrong query is given example a string
> * with respect to above, write test for wrong query parameters
> * work on the changes observed by Adeola especially the location of your route

Lundii commented on test/server/routes/auth.test.js:
```diff
@@ -406,6 +407,27 @@ describe('AUTH', () => {
+  describe('Pagination router', () => {
+    it('It should get book with query string page and limit', (done) => {
+      chai.request(server)
+        .get(`${BOOKS_BASE_URL}?page=1&limit=10`)
```
I saw you added a validator which is great but what I was talking about is a validation that checks the query parameters and test for them. For example, what happens if I tried to access the url by writing the following
* `/api/v1/books?page=today&limit=tomorrow` or
* just the limit parameter like `api/v1/books?limit=10` and similar queries,

Nennyfills replied:
![image](https://user-images.githubusercontent.com/45515619/59558042-bc883480-8fe0-11e9-856c-f05635ff9d29.png)

Nennyfills replied:
please, can we try to run the server locally before leaving a comment? because if you did this comment would have been prevented.

Lundii replied:
Alright

Adekoreday reviewed (commented):
Line 67 book controller successful was  not well spelt

Ukhu reviewed (changes requested):
From your screenshot in the PR description. The books returned by the server do not contain the name of the author that wrote the book.

Nennyfills commented:
> From your screenshot in the PR description. The books returned by the server do not contain the name of the author that wrote the book.

from the pivotal story, am working on pagination,  showing the page and the limit.

Nennyfills commented:
> > From your screenshot in the PR description. The books returned by the server do not contain the name of the author that wrote the book.
> 
> from the pivotal story, am working on pagination, showing the page and the limit.

Done

Ukhu reviewed (changes requested):
Well done on what you have done so far, however, these are my observations after reviewing;

1) I noticed that there are 2 seeder files for authors. It shouldn't be so.

2) From the response I got from the server when I tested this route, some books didn't have any authors. All books in the database should have authors.
![Screenshot (170)](https://user-images.githubusercontent.com/35268825/59642918-c768d380-915e-11e9-92b6-2ea5b1c808e6.png)

3) It does not handle situations where a user inputs an alphabet or any other non-integer in the query params for page or limit. You should also write test cases for these.

4) The results in the server response were not ordered. Book with ID of 10 came before book with ID of 8, and book with ID of 9 came before book with ID of 7.

Recommendations:

For the ordering, you can use Sequelize's _order_ option to order by ID in ascending order. You can also associate the books an authors through the BookAuthor table so that all the books can have authors.

Nennyfills commented:
> Well done on what you have done so far, however, these are my observations after reviewing;
> 
> 1. I noticed that there are 2 seeder files for authors. It shouldn't be so.
> 2. From the response I got from the server when I tested this route, some books didn't have any authors. All books in the database should have authors.
>    ![Screenshot (170)](https://user-images.githubusercontent.com/35268825/59642918-c768d380-915e-11e9-92b6-2ea5b1c808e6.png)
> 3. It does not handle situations where a user inputs an alphabet or any other non-integer in the query params for page or limit. You should also write test cases for these.
> 4. The results in the server response were not ordered. Book with ID of 10 came before book with ID of 8, and book with ID of 9 came before book with ID of 7.
> 
> Recommendations:
> 
> For the ordering, you can use Sequelize's _order_ option to order by ID in ascending order. You can also associate the books an authors through the BookAuthor table so that all the books can have authors.

Fixed all changes, for the response not ordered, really can't tell why yours wasn't ordered, mine was ordered, but for satisfactory purpose, I have added the order option to it.
![image](https://user-images.githubusercontent.com/45515619/59678771-25300680-91c5-11e9-8369-4f3ac497d46c.png)

Ukhu commented on server/middlewares/BookValidator.js:
```diff
@@ -126,6 +140,20 @@ export default class BookValidator {
+      BookValidator.checkNumberForDefault('page'),
+      BookValidator.checkNumberForDefault('limit'),
+      checkForErrors,
+      checkForErrors,
```
Why are there two 'checkForErrors' middlewares?

Nennyfills replied:
done

Ukhu commented on test/server/routes/auth.test.js:
```diff
@@ -406,6 +407,37 @@ describe('AUTH', () => {
+          expect(response.body.message).to.equal('no more books');
+          done();
+        });
+    });
```
There are no test cases for when a user tries to input wrong data types e.g alphabets or special characters in the parameter field for page or limit.

Nennyfills replied:
done

Nennyfills committed:
166214113-story(users): pagination for book
- create new endpoints to fetchBooks  for all users
- add faker library to the seeders book file
- write unit test to test all routers
 [Started  #166214113]
